File: VideoStreamingBackend/src/infrastructure/streaming/StreamManager.ts
/**
 * StreamManager — Orchestrates camera ingest and viewer lifecycle.
 *
 * Supports DUAL mode:
 *   1. JPEG over WebSocket (legacy frontend) — FFmpegProcess → binary frames
 *   2. WebRTC via SFU (new path)             — MediasoupSFU → mediasoup consumers
 *
 * Both pipelines run simultaneously per camera so old and new clients coexist.
 *
 * Architecture:
 *   StreamManager uses ISFUService (port) → MediasoupSFU (adapter)
 *   StreamManager uses FFmpegProcess for legacy JPEG extraction
 *   StreamManager uses IStreamSessionRepository for persistence
 *   Application layer use cases call StreamManager methods
 */

import { EventEmitter } from 'events';
import { v4 as uuidv4 } from 'uuid';
import { StreamSession, StreamStatus, DEFAULT_STREAM_CONFIG, StreamConfig } from '../../domain/entities/StreamSession';
import { IStreamSessionRepository } from '../../domain/repositories';
import {
  ISFUService,
  RtpCapabilities,
  WebRtcTransportParams,
  DtlsParameters,
  ConsumerParams,
} from '../../domain/services';
import { AudioExtractor, FFmpegProcess, probeStream } from './FFmpegProcess';
import { config } from '../config';
import { logger } from '../logging/logger';

export class StreamManager extends EventEmitter {
  private sessions = new Map<string, StreamSession>();

  /** FFmpeg JPEG extraction processes (legacy JPEG WS path) */
  private jpegProcesses = new Map<string, FFmpegProcess>();
  /** Latest JPEG frame per camera (for HTTP polling fallback) */
  private latestFrames = new Map<string, Buffer>();

  /** FFmpeg audio extraction per camera (raw PCM over WS) */
  private audioExtractors = new Map<string, AudioExtractor>();
  private audioViewerCounts = new Map<string, number>();
  private audioInfo = new Map<string, { sampleRate: number; channels: number; hasAudio: boolean }>();
  private audioStopTimers = new Map<string, NodeJS.Timeout>();

  constructor(
    private readonly repo: IStreamSessionRepository,
    private readonly sfu: ISFUService | null,
  ) {
    super();
  }

  private async getOrProbeAudioInfo(cameraId: string): Promise<{ sampleRate: number; channels: number; hasAudio: boolean }> {
    const existing = this.audioInfo.get(cameraId);
    if (existing) return existing;

    const session = this.sessions.get(cameraId);
    if (!session) {
      const fallback = { hasAudio: false, sampleRate: 48000, channels: 1 };
      this.audioInfo.set(cameraId, fallback);
      return fallback;
    }

    try {
      const probe = await probeStream(session.streamUrl);
      const info = {
        hasAudio: probe.hasAudio,
        sampleRate: probe.audioSampleRate ?? 48000,
        channels: probe.audioChannels ?? 1,
      };
      this.audioInfo.set(cameraId, info);
      return info;
    } catch {
      const fallback = { hasAudio: false, sampleRate: 48000, channels: 1 };
      this.audioInfo.set(cameraId, fallback);
      return fallback;
    }
  }

  /**
   * Start audio extraction if needed and increment audio viewer count.
   * Returns negotiated audio metadata for the frontend.
   */
  async addAudioViewer(cameraId: string): Promise<{ sampleRate: number; channels: number; hasAudio: boolean }> {
    const current = this.audioViewerCounts.get(cameraId) ?? 0;
    this.audioViewerCounts.set(cameraId, current + 1);

    const timer = this.audioStopTimers.get(cameraId);
    if (timer) {
      clearTimeout(timer);
      this.audioStopTimers.delete(cameraId);
    }

    const info = await this.getOrProbeAudioInfo(cameraId);
    this.ensureAudioExtraction(cameraId, info);
    return info;
  }

  removeAudioViewer(cameraId: string): void {
    const current = this.audioViewerCounts.get(cameraId) ?? 0;
    const next = Math.max(0, current - 1);
    this.audioViewerCounts.set(cameraId, next);

    if (next > 0) return;
    if (this.audioStopTimers.has(cameraId)) return;

    const timeout = setTimeout(() => {
      this.stopAudioExtraction(cameraId);
    }, 5000);
    timeout.unref();
    this.audioStopTimers.set(cameraId, timeout);
  }

  // ═══════════════════════════════════════════════════════════════
  // Camera Ingest (RTSP → SFU)
  // ═══════════════════════════════════════════════════════════════

  /**
   * Start streaming for a camera.
   * Launches JPEG extraction (always) and SFU ingest (if available).
   */
  async startStream(
    cameraId: string,
    ownerUserId: string,
    streamUrl: string,
    partialConfig?: Partial<StreamConfig>,
  ): Promise<StreamSession> {
    // Return existing if already active
    const existing = this.sessions.get(cameraId);
    if (existing && (existing.status === 'active' || existing.status === 'connecting')) {
      return existing;
    }

    const streamConfig: StreamConfig = {
      ...DEFAULT_STREAM_CONFIG,
      fps: partialConfig?.fps ?? config.defaultFps,
      width: partialConfig?.width ?? config.defaultWidth,
      height: partialConfig?.height ?? config.defaultHeight,
      codec: partialConfig?.codec ?? DEFAULT_STREAM_CONFIG.codec,
      bitrate: partialConfig?.bitrate ?? DEFAULT_STREAM_CONFIG.bitrate,
    };

    const session: StreamSession = {
      id: uuidv4(),
      cameraId,
      ownerUserId,
      streamUrl,
      status: 'connecting',
      config: streamConfig,
      viewerCount: 0,
      routerId: null,
      videoProducerId: null,
      audioProducerId: null,
      startedAt: null,
      lastFrameAt: null,
      stoppedAt: null,
      errorMessage: null,
      reconnectAttempts: 0,
      createdAt: new Date(),
      updatedAt: new Date(),
    };

    this.sessions.set(cameraId, session);
    this.repo.save(session);

    // ── Always start JPEG extraction for legacy WS viewers ──
    this.startJpegExtraction(cameraId, streamUrl, streamConfig);

    // ── Start audio extraction lazily (only if an audio viewer connects) ──
    // Audio extraction is triggered via addAudioViewer() so we don't waste CPU.

    // ── Try SFU ingest (if mediasoup is available) ──
    if (this.sfu) {
      try {
        const result = await this.sfu.ingestCamera(cameraId, streamUrl, {
          fps: streamConfig.fps,
          width: streamConfig.width,
          height: streamConfig.height,
        });

        session.routerId = result.routerId;
        session.videoProducerId = result.videoProducerId;
        session.audioProducerId = result.audioProducerId;
      } catch (err: any) {
        logger.warn(`[StreamManager] SFU ingest failed for ${cameraId} (JPEG still active): ${err.message}`);
      }
    }

    session.status = 'active';
    session.startedAt = new Date();
    session.updatedAt = new Date();
    this.repo.save(session);

    logger.info(`[StreamManager] Stream active for camera ${cameraId}`);
    return session;
  }

  private ensureAudioExtraction(cameraId: string, info: { sampleRate: number; channels: number; hasAudio: boolean }): void {
    if (this.audioExtractors.has(cameraId)) return;
    const session = this.sessions.get(cameraId);
    if (!session) return;
    if (!info.hasAudio) return;

    // If no audio viewers, don't start.
    const viewers = this.audioViewerCounts.get(cameraId) ?? 0;
    if (viewers <= 0) return;

    const extractor = new AudioExtractor();
    extractor.on('audio', (chunk: Buffer) => {
      this.emit('audio', cameraId, chunk);
    });
    extractor.on('close', (code: number) => {
      logger.info(`[StreamManager] Audio FFmpeg exited for ${cameraId} (code ${code})`);
      this.audioExtractors.delete(cameraId);
    });
    extractor.on('error', (err: Error) => {
      logger.warn(`[StreamManager] Audio FFmpeg error for ${cameraId}: ${err.message}`);
      this.audioExtractors.delete(cameraId);
    });

    this.audioExtractors.set(cameraId, extractor);
    extractor.start(session.streamUrl, info.sampleRate, info.channels);
    logger.info(`[StreamManager] Audio extraction started for ${cameraId}`);
  }

  private stopAudioExtraction(cameraId: string): void {
    const proc = this.audioExtractors.get(cameraId);
    if (proc) {
      proc.stop();
      this.audioExtractors.delete(cameraId);
    }
  }

  // ─── JPEG Frame Extraction (legacy WS + HTTP polling) ────────

  private startJpegExtraction(cameraId: string, streamUrl: string, streamConfig: StreamConfig): void {
    if (this.jpegProcesses.has(cameraId)) return;

    const ffmpeg = new FFmpegProcess();

    ffmpeg.on('frame', (frame: Buffer) => {
      this.latestFrames.set(cameraId, frame);
      // Update lastFrameAt
      const session = this.sessions.get(cameraId);
      if (session) {
        session.lastFrameAt = new Date();
      }
      // Fan out to WS viewers listening on the 'frame' event
      this.emit('frame', cameraId, frame);
    });

    ffmpeg.on('close', (code: number) => {
      logger.info(`[StreamManager] JPEG FFmpeg exited for ${cameraId} (code ${code})`);
      this.jpegProcesses.delete(cameraId);

      // Auto-reconnect if session is still meant to be active
      const session = this.sessions.get(cameraId);
      if (session && session.status === 'active' && session.reconnectAttempts < config.maxReconnectAttempts) {
        session.status = 'reconnecting';
        session.reconnectAttempts++;
        session.updatedAt = new Date();
        logger.info(`[StreamManager] Reconnecting JPEG for ${cameraId} (attempt ${session.reconnectAttempts})`);
        setTimeout(() => {
          if (this.sessions.has(cameraId)) {
            this.startJpegExtraction(cameraId, streamUrl, streamConfig);
            const s = this.sessions.get(cameraId);
            if (s) { s.status = 'active'; s.updatedAt = new Date(); }
          }
        }, config.reconnectDelayMs);
      }
    });

    ffmpeg.on('error', (err: Error) => {
      logger.error(`[StreamManager] JPEG FFmpeg error for ${cameraId}: ${err.message}`);
    });

    ffmpeg.start({
      streamUrl,
      fps: streamConfig.fps,
      width: streamConfig.width,
      height: streamConfig.height,
    });

    this.jpegProcesses.set(cameraId, ffmpeg);
    logger.info(`[StreamManager] JPEG extraction started for ${cameraId}`);
  }

  /** Get latest JPEG frame for HTTP polling */
  getLatestFrame(cameraId: string): Buffer | null {
    return this.latestFrames.get(cameraId) ?? null;
  }

  /**
   * Stop streaming for a camera (both JPEG and SFU).
   */
  stopStream(cameraId: string): StreamSession | null {
    const session = this.sessions.get(cameraId);
    if (!session) return null;

    // Stop JPEG extraction
    const ffmpeg = this.jpegProcesses.get(cameraId);
    if (ffmpeg) {
      ffmpeg.stop();
      this.jpegProcesses.delete(cameraId);
    }
    this.latestFrames.delete(cameraId);

    // Stop audio extraction
    this.stopAudioExtraction(cameraId);
    this.audioInfo.delete(cameraId);

    // Stop SFU ingest
    if (this.sfu) {
      this.sfu.stopIngest(cameraId);
    }

    session.status = 'stopped';
    session.stoppedAt = new Date();
    session.updatedAt = new Date();
    this.repo.save(session);
    this.sessions.delete(cameraId);

    logger.info(`[StreamManager] Stopped stream for camera ${cameraId}`);
    return session;
  }

  stopAll(): void {
    for (const cameraId of Array.from(this.sessions.keys())) {
      this.stopStream(cameraId);
    }
  }

  // ═══════════════════════════════════════════════════════════════
  // WebRTC Viewer Negotiation (Signaling support)
  // ═══════════════════════════════════════════════════════════════

  /** Get Router RTP capabilities for a camera (needed by browser Device) */
  getRouterRtpCapabilities(cameraId: string): RtpCapabilities | null {
    return this.sfu?.getRouterRtpCapabilities(cameraId) ?? null;
  }

  /** Create a WebRtcTransport for a viewer */
  async createViewerTransport(cameraId: string): Promise<WebRtcTransportParams> {
    if (!this.sfu) throw new Error('SFU not available');
    return this.sfu.createViewerTransport(cameraId);
  }

  /** Connect a viewer's transport (ICE + DTLS) */
  async connectViewerTransport(cameraId: string, transportId: string, dtlsParameters: DtlsParameters): Promise<void> {
    if (!this.sfu) throw new Error('SFU not available');
    return this.sfu.connectViewerTransport(cameraId, transportId, dtlsParameters);
  }

  /** Create a Consumer for a viewer */
  async createConsumer(
    cameraId: string,
    transportId: string,
    producerId: string,
    rtpCapabilities: RtpCapabilities,
  ): Promise<ConsumerParams> {
    if (!this.sfu) throw new Error('SFU not available');
    return this.sfu.createConsumer(cameraId, transportId, producerId, rtpCapabilities);
  }

  /** Resume a Consumer */
  async resumeConsumer(cameraId: string, consumerId: string): Promise<void> {
    if (!this.sfu) throw new Error('SFU not available');
    return this.sfu.resumeConsumer(cameraId, consumerId);
  }

  /** Close a viewer's transport */
  closeViewerTransport(cameraId: string, transportId: string): void {
    this.sfu?.closeViewerTransport(cameraId, transportId);
  }

  // ═══════════════════════════════════════════════════════════════
  // Session Queries
  // ═══════════════════════════════════════════════════════════════

  getSession(cameraId: string): StreamSession | undefined {
    return this.sessions.get(cameraId);
  }

  getAllSessions(): StreamSession[] {
    return Array.from(this.sessions.values());
  }

  isStreaming(cameraId: string): boolean {
    const s = this.sessions.get(cameraId);
    return !!s && (s.status === 'active' || s.status === 'connecting');
  }

  /**
   * Get real-time info for a streaming camera.
   * Tries SFU stats first, falls back to JPEG pipeline info.
   */
  async getRealTimeInfo(cameraId: string): Promise<{
    currentFps: number;
    viewerCount: number;
    hasAudio: boolean;
    status: StreamStatus;
    uptime: number;
    bitrate: number;
  } | null> {
    const session = this.sessions.get(cameraId);
    if (!session) return null;

    let currentFps = 0;
    let hasAudio = false;
    let bitrate = 0;

    // Try SFU stats
    if (this.sfu) {
      try {
        const stats = await this.sfu.getStats(cameraId);
        if (stats) {
          currentFps = stats.currentFps ?? 0;
          hasAudio = stats.hasAudio ?? false;
          bitrate = stats.bitrate ?? 0;
        }
      } catch { /* SFU stats not available */ }
    }

    // Estimate FPS from JPEG frames if SFU stats unavailable
    if (currentFps === 0 && session.config) {
      currentFps = this.jpegProcesses.has(cameraId) ? session.config.fps : 0;
    }

    return {
      currentFps,
      viewerCount: session.viewerCount,
      hasAudio,
      status: session.status,
      uptime: session.startedAt
        ? Math.floor((Date.now() - session.startedAt.getTime()) / 1000)
        : 0,
      bitrate,
    };
  }

  // ═══════════════════════════════════════════════════════════════
  // Viewer counting
  // ═══════════════════════════════════════════════════════════════

  addViewer(cameraId: string): void {
    const s = this.sessions.get(cameraId);
    if (s) {
      s.viewerCount++;
      s.updatedAt = new Date();
    }
  }

  removeViewer(cameraId: string): void {
    const s = this.sessions.get(cameraId);
    if (s && s.viewerCount > 0) {
      s.viewerCount--;
      s.updatedAt = new Date();
    }
  }

  // ═══════════════════════════════════════════════════════════════
  // ICE server config for clients
  // ═══════════════════════════════════════════════════════════════

  getIceServers(): Array<{ urls: string; username?: string; credential?: string }> {
    const servers: Array<{ urls: string; username?: string; credential?: string }> = [];

    if (config.stunServer) {
      servers.push({ urls: config.stunServer });
    }

    if (config.turnServer && config.turnUsername && config.turnCredential) {
      servers.push({
        urls: config.turnServer,
        username: config.turnUsername,
        credential: config.turnCredential,
      });
    }

    return servers;
  }
}

File: StreamingBackend/internal/infrastructure/streaming/stream_manager.go
package streaming

import (
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/pion/webrtc/v4"
	"github.com/rs/zerolog/log"
	"github.com/vigileye/streaming-backend/internal/domain/entities"
	domainerrors "github.com/vigileye/streaming-backend/internal/domain/errors"
	"github.com/vigileye/streaming-backend/internal/domain/repositories"
	"github.com/vigileye/streaming-backend/internal/domain/services"
	"github.com/vigileye/streaming-backend/internal/infrastructure/config"
)

// Ensure StreamManager implements services.StreamService
var _ services.StreamService = (*StreamManager)(nil)

// StreamManager orchestrates camera ingest and viewer lifecycle.
// When MediaMTX is enabled, it delegates stream ingestion to MediaMTX
// and proxies WHEP signaling for WebRTC viewers.
type StreamManager struct {
	mu sync.RWMutex

	cfg      *config.Config
	repo     repositories.StreamSessionRepository

	// Legacy pion-based ingests (used when MediaMTX is disabled)
	ingests  map[string]*CameraIngest

	// MediaMTX client (nil when MediaMTX is disabled)
	mediaMTX *MediaMTXClient

	// WHEP session URLs: viewerID → MediaMTX session URL
	whepSessions map[string]string
	whepMu       sync.RWMutex

	// JPEG frames for HTTP polling fallback
	jpegProcesses map[string]*FFmpegProcess
	latestFrames  map[string][]byte
	frameMu       sync.RWMutex
}

// NewStreamManager creates a new stream manager.
func NewStreamManager(cfg *config.Config, repo repositories.StreamSessionRepository) *StreamManager {
	sm := &StreamManager{
		cfg:           cfg,
		repo:          repo,
		ingests:       make(map[string]*CameraIngest),
		whepSessions:  make(map[string]string),
		jpegProcesses: make(map[string]*FFmpegProcess),
		latestFrames:  make(map[string][]byte),
	}

	if cfg.MediaMTXEnabled {
		sm.mediaMTX = NewMediaMTXClient(cfg.MediaMTXAPIURL)
		// Override WHEP base URL if the WHEP URL is different from API URL
		if cfg.MediaMTXWHEPURL != "" && cfg.MediaMTXWHEPURL != cfg.MediaMTXAPIURL {
			sm.mediaMTX = NewMediaMTXClient(cfg.MediaMTXAPIURL)
		}
		log.Info().
			Str("api_url", cfg.MediaMTXAPIURL).
			Str("whep_url", cfg.MediaMTXWHEPURL).
			Msg("MediaMTX integration enabled")
	} else {
		log.Info().Msg("MediaMTX disabled, using legacy pion/FFmpeg pipeline")
	}

	return sm
}

// GetMediaMTXClient returns the MediaMTX client (may be nil).
func (sm *StreamManager) GetMediaMTXClient() *MediaMTXClient {
	return sm.mediaMTX
}

// IsMediaMTXEnabled returns whether MediaMTX mode is active.
func (sm *StreamManager) IsMediaMTXEnabled() bool {
	return sm.cfg.MediaMTXEnabled && sm.mediaMTX != nil
}

// StartStream begins streaming a camera.
// With MediaMTX: registers the camera's stream URL as a MediaMTX path.
// Without MediaMTX: uses FFmpeg → RTP → pion TrackLocalStaticRTP pipeline.
func (sm *StreamManager) StartStream(cameraID, ownerUserID, streamURL string, cfgOverride *entities.StreamConfig) (*entities.StreamSession, error) {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	// Return existing if already active
	if existing, ok := sm.repo.Get(cameraID); ok && existing.IsActive() {
		return existing, nil
	}

	streamCfg := entities.DefaultStreamConfig
	if cfgOverride != nil {
		if cfgOverride.FPS > 0 {
			streamCfg.FPS = cfgOverride.FPS
		}
		if cfgOverride.Width > 0 {
			streamCfg.Width = cfgOverride.Width
		}
		if cfgOverride.Height > 0 {
			streamCfg.Height = cfgOverride.Height
		}
		if cfgOverride.Codec != "" {
			streamCfg.Codec = cfgOverride.Codec
		}
		if cfgOverride.Bitrate > 0 {
			streamCfg.Bitrate = cfgOverride.Bitrate
		}
	}

	session := entities.NewStreamSession(
		uuid.New().String(),
		cameraID,
		ownerUserID,
		streamURL,
		streamCfg,
	)
	sm.repo.Save(session)

	if sm.IsMediaMTXEnabled() {
		// ─── MediaMTX path ───
		if err := sm.mediaMTX.AddPath(cameraID, streamURL); err != nil {
			log.Error().Err(err).Str("camera_id", cameraID).Msg("Failed to add MediaMTX path")
			session.SetError("MediaMTX path creation failed: " + err.Error())
			sm.repo.Save(session)
			return nil, domainerrors.NewIngestFailedError(cameraID, err.Error())
		}
		log.Info().Str("camera_id", cameraID).Str("source", streamURL).Msg("MediaMTX path registered")
	} else {
		// ─── Legacy: FFmpeg → RTP → pion ───
		ingest := NewCameraIngest(cameraID, streamURL, streamCfg.FPS, streamCfg.Width, streamCfg.Height, sm.cfg)
		if err := ingest.Start(); err != nil {
			log.Warn().Err(err).Str("camera_id", cameraID).Msg("WebRTC ingest failed, running JPEG-only")
		} else {
			sm.ingests[cameraID] = ingest
		}
	}

	// Start JPEG extraction for HTTP polling fallback (both modes)
	sm.startJPEGExtraction(cameraID, streamURL, streamCfg)

	session.Activate()
	sm.repo.Save(session)

	log.Info().
		Str("camera_id", cameraID).
		Str("stream_url", streamURL).
		Int("fps", streamCfg.FPS).
		Bool("mediamtx", sm.IsMediaMTXEnabled()).
		Msg("Stream started")

	return session, nil
}

// StopStream stops a camera stream and all its viewers.
func (sm *StreamManager) StopStream(cameraID string) (*entities.StreamSession, error) {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	session, ok := sm.repo.Get(cameraID)
	if !ok {
		return nil, domainerrors.NewStreamNotFoundError(cameraID)
	}

	if sm.IsMediaMTXEnabled() {
		// Remove path from MediaMTX
		if err := sm.mediaMTX.RemovePath(cameraID); err != nil {
			log.Warn().Err(err).Str("camera_id", cameraID).Msg("Failed to remove MediaMTX path")
		}
	} else {
		// Stop legacy WebRTC ingest
		if ingest, ok := sm.ingests[cameraID]; ok {
			ingest.Stop()
			delete(sm.ingests, cameraID)
		}
	}

	// Stop JPEG extraction
	sm.stopJPEGExtraction(cameraID)

	// Clean up WHEP sessions for this camera
	sm.cleanWHEPSessionsForCamera(cameraID)

	session.Stop()
	sm.repo.Save(session)
	sm.repo.Remove(cameraID)

	log.Info().Str("camera_id", cameraID).Msg("Stream stopped")
	return session, nil
}

// StopAll stops all active streams.
func (sm *StreamManager) StopAll() {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if sm.IsMediaMTXEnabled() {
		// Remove all paths from MediaMTX
		for _, session := range sm.repo.GetAll() {
			if err := sm.mediaMTX.RemovePath(session.CameraID); err != nil {
				log.Warn().Err(err).Str("camera_id", session.CameraID).Msg("Failed to remove MediaMTX path")
			}
		}
	} else {
		for cameraID, ingest := range sm.ingests {
			ingest.Stop()
			delete(sm.ingests, cameraID)
		}
	}

	for cameraID := range sm.jpegProcesses {
		sm.stopJPEGExtraction(cameraID)
	}

	log.Info().Msg("All streams stopped")
}

// GetSession returns the session for a camera.
func (sm *StreamManager) GetSession(cameraID string) (*entities.StreamSession, bool) {
	return sm.repo.Get(cameraID)
}

// GetAllSessions returns all sessions.
func (sm *StreamManager) GetAllSessions() []*entities.StreamSession {
	return sm.repo.GetAll()
}

// IsStreaming checks if a camera is actively streaming.
func (sm *StreamManager) IsStreaming(cameraID string) bool {
	session, ok := sm.repo.Get(cameraID)
	return ok && session.IsActive()
}

// AddViewer increments the viewer count.
func (sm *StreamManager) AddViewer(cameraID string) {
	if session, ok := sm.repo.Get(cameraID); ok {
		session.IncrementViewers()
		sm.repo.Save(session)
	}
}

// RemoveViewer decrements the viewer count.
func (sm *StreamManager) RemoveViewer(cameraID string) {
	if session, ok := sm.repo.Get(cameraID); ok {
		session.DecrementViewers()
		sm.repo.Save(session)
	}
}

// GetRealTimeInfo returns streaming statistics.
func (sm *StreamManager) GetRealTimeInfo(cameraID string) *services.RealTimeInfo {
	session, ok := sm.repo.Get(cameraID)
	if !ok {
		return nil
	}

	info := &services.RealTimeInfo{
		Status:      string(session.Status),
		ViewerCount: session.ViewerCount,
	}

	if sm.IsMediaMTXEnabled() {
		// Get stats from MediaMTX
		pathStatus, err := sm.mediaMTX.GetPathStatus(cameraID)
		if err == nil && pathStatus != nil {
			info.ViewerCount = len(pathStatus.Readers)
			if pathStatus.Ready {
				info.Status = "active"
			}
		}
		info.CurrentFPS = session.Config.FPS
	} else {
		sm.mu.RLock()
		ingest, hasIngest := sm.ingests[cameraID]
		sm.mu.RUnlock()

		if hasIngest {
			fps, bitrate, viewers := ingest.GetStats()
			info.CurrentFPS = fps
			info.Bitrate = bitrate
			if viewers > info.ViewerCount {
				info.ViewerCount = viewers
			}
		} else if session.IsActive() {
			info.CurrentFPS = session.Config.FPS
		}
	}

	if session.StartedAt != nil {
		info.Uptime = int64(time.Since(*session.StartedAt).Seconds())
	}

	return info
}

// GetICEServers returns ICE server configuration.
func (sm *StreamManager) GetICEServers() []services.ICEServer {
	servers := []services.ICEServer{}

	if sm.cfg.STUNServer != "" {
		servers = append(servers, services.ICEServer{URLs: sm.cfg.STUNServer})
	}
	if sm.cfg.TURNServer != "" && sm.cfg.TURNUsername != "" && sm.cfg.TURNCredential != "" {
		servers = append(servers, services.ICEServer{
			URLs:       sm.cfg.TURNServer,
			Username:   sm.cfg.TURNUsername,
			Credential: sm.cfg.TURNCredential,
		})
	}

	return servers
}

// GetLatestFrame returns the latest JPEG frame for HTTP polling.
func (sm *StreamManager) GetLatestFrame(cameraID string) []byte {
	sm.frameMu.RLock()
	defer sm.frameMu.RUnlock()
	return sm.latestFrames[cameraID]
}

// ─── WebRTC Signaling ───

// NegotiateWebRTC handles a WebRTC offer from a viewer and returns the answer.
// With MediaMTX: proxies the SDP offer/answer via WHEP.
// Without MediaMTX: uses pion PeerConnection directly.
func (sm *StreamManager) NegotiateWebRTC(cameraID, viewerID string, offer webrtc.SessionDescription) (*webrtc.SessionDescription, error) {
	if sm.IsMediaMTXEnabled() {
		return sm.negotiateViaWHEP(cameraID, viewerID, offer)
	}
	return sm.negotiateViaLegacy(cameraID, viewerID, offer)
}

// negotiateViaWHEP proxies WebRTC signaling through MediaMTX WHEP.
func (sm *StreamManager) negotiateViaWHEP(cameraID, viewerID string, offer webrtc.SessionDescription) (*webrtc.SessionDescription, error) {
	// Use the WHEP-specific URL if configured, otherwise fall back to API URL
	whepClient := NewMediaMTXClient(sm.cfg.MediaMTXWHEPURL)

	answerSDP, sessionURL, err := whepClient.WHEPOffer(cameraID, offer.SDP)
	if err != nil {
		return nil, domainerrors.NewStreamConnectionError(cameraID, "WHEP negotiation failed: "+err.Error())
	}

	// Store WHEP session URL for this viewer (for ICE candidates and cleanup)
	sm.whepMu.Lock()
	sm.whepSessions[viewerID] = sessionURL
	sm.whepMu.Unlock()

	answer := &webrtc.SessionDescription{
		Type: webrtc.SDPTypeAnswer,
		SDP:  answerSDP,
	}

	log.Info().
		Str("camera_id", cameraID).
		Str("viewer_id", viewerID).
		Str("session_url", sessionURL).
		Msg("WHEP session established via MediaMTX")

	return answer, nil
}

// negotiateViaLegacy uses the pion-based direct WebRTC pipeline.
func (sm *StreamManager) negotiateViaLegacy(cameraID, viewerID string, offer webrtc.SessionDescription) (*webrtc.SessionDescription, error) {
	sm.mu.RLock()
	ingest, ok := sm.ingests[cameraID]
	sm.mu.RUnlock()

	if !ok {
		return nil, domainerrors.NewStreamNotFoundError(cameraID)
	}

	return ingest.AddViewer(viewerID, offer)
}

// AddICECandidate adds an ICE candidate for a viewer.
func (sm *StreamManager) AddICECandidate(cameraID, viewerID string, candidate webrtc.ICECandidateInit) error {
	if sm.IsMediaMTXEnabled() {
		return sm.addICEViaWHEP(viewerID, candidate)
	}

	sm.mu.RLock()
	ingest, ok := sm.ingests[cameraID]
	sm.mu.RUnlock()

	if !ok {
		return domainerrors.NewStreamNotFoundError(cameraID)
	}

	return ingest.AddICECandidate(viewerID, candidate)
}

// addICEViaWHEP sends a trickle ICE candidate to MediaMTX via WHEP.
func (sm *StreamManager) addICEViaWHEP(viewerID string, candidate webrtc.ICECandidateInit) error {
	sm.whepMu.RLock()
	sessionURL, ok := sm.whepSessions[viewerID]
	sm.whepMu.RUnlock()

	if !ok {
		return domainerrors.NewValidationError("WHEP session not found for viewer: " + viewerID)
	}

	whepClient := NewMediaMTXClient(sm.cfg.MediaMTXWHEPURL)
	return whepClient.WHEPAddICECandidate(sessionURL, candidate.Candidate)
}

// DisconnectViewer removes a viewer from a camera's ingest.
func (sm *StreamManager) DisconnectViewer(cameraID, viewerID string) {
	if sm.IsMediaMTXEnabled() {
		sm.disconnectViaWHEP(viewerID)
	} else {
		sm.mu.RLock()
		ingest, ok := sm.ingests[cameraID]
		sm.mu.RUnlock()

		if ok {
			ingest.RemoveViewer(viewerID)
		}
	}
	sm.RemoveViewer(cameraID)
}

// disconnectViaWHEP tears down a WHEP session.
func (sm *StreamManager) disconnectViaWHEP(viewerID string) {
	sm.whepMu.Lock()
	sessionURL, ok := sm.whepSessions[viewerID]
	if ok {
		delete(sm.whepSessions, viewerID)
	}
	sm.whepMu.Unlock()

	if ok && sessionURL != "" {
		whepClient := NewMediaMTXClient(sm.cfg.MediaMTXWHEPURL)
		if err := whepClient.WHEPDelete(sessionURL); err != nil {
			log.Warn().Err(err).Str("viewer_id", viewerID).Msg("Failed to delete WHEP session")
		}
	}
}

// cleanWHEPSessionsForCamera removes all WHEP sessions that belong to a camera.
// Note: we don't track camera→viewer mapping for WHEP, but we clean all for simplicity.
func (sm *StreamManager) cleanWHEPSessionsForCamera(cameraID string) {
	// In a production system, you'd maintain a camera→viewerIDs index.
	// For now, WHEP sessions will time out in MediaMTX automatically.
	_ = cameraID
}

// ─── JPEG Extraction (HTTP polling fallback) ───

func (sm *StreamManager) startJPEGExtraction(cameraID, streamURL string, cfg entities.StreamConfig) {
	if _, ok := sm.jpegProcesses[cameraID]; ok {
		return
	}

	ffmpeg := NewFFmpegProcess()
	ffmpeg.OnFrame = func(frame []byte) {
		sm.frameMu.Lock()
		sm.latestFrames[cameraID] = frame
		sm.frameMu.Unlock()

		if session, ok := sm.repo.Get(cameraID); ok {
			now := time.Now().UTC()
			session.LastFrameAt = &now
		}
	}

	ffmpeg.OnClose = func(code int) {
		log.Info().Str("camera_id", cameraID).Int("exit_code", code).Msg("JPEG FFmpeg exited")
		delete(sm.jpegProcesses, cameraID)

		// Auto-reconnect
		if session, ok := sm.repo.Get(cameraID); ok && session.IsActive() && session.ReconnectAttempts < sm.cfg.MaxReconnectAttempts {
			session.ReconnectAttempts++
			time.AfterFunc(time.Duration(sm.cfg.ReconnectDelayMs)*time.Millisecond, func() {
				sm.mu.Lock()
				defer sm.mu.Unlock()
				if _, stillActive := sm.repo.Get(cameraID); stillActive {
					sm.startJPEGExtraction(cameraID, streamURL, cfg)
				}
			})
		}
	}

	if err := ffmpeg.Start(FFmpegOptions{
		StreamURL: streamURL,
		FPS:       cfg.FPS,
		Width:     cfg.Width,
		Height:    cfg.Height,
		FFmpegBin: sm.cfg.FFmpegPath,
	}); err != nil {
		log.Error().Err(err).Str("camera_id", cameraID).Msg("Failed to start JPEG extraction")
		return
	}

	sm.jpegProcesses[cameraID] = ffmpeg
}

func (sm *StreamManager) stopJPEGExtraction(cameraID string) {
	if ffmpeg, ok := sm.jpegProcesses[cameraID]; ok {
		ffmpeg.Stop()
		delete(sm.jpegProcesses, cameraID)
	}
	sm.frameMu.Lock()
	delete(sm.latestFrames, cameraID)
	sm.frameMu.Unlock()
}
